feat(caching): tie lifetime of cached tables to python refs

coroa opened the pull request:
## Description of changes

Makes use of `weakref.finalizer`s to release the cached table automatically when the cached table (or more precisely its `.op()`) is garbage collected.

```python
In [1]: import pandas as pd

In [2]: import ibis

In [3]: con = ibis.duckdb.connect()

In [4]: df = pd.DataFrame([[0, 1], [2, 3]], columns=["foo", "bar"])

In [5]: noncached = con.create_table("tab", df).mutate(foo=2 * tab.foo)

In [6]: cached = noncached.cache()

In [7]: con.tables
Out[7]: 
Tables
------
- ibis_cache_wvrjoptclzhudglqapz4zusnlu
- tab

In [8]: del cached

In [9]: con.tables
Out[9]: 
Tables
------
- tab
```

If I understand the nodal expression tree structure correctly, then `cached.op()` is preserved in all dependent expressions. This PR adds a `weakref.finalize(cached.op(), ...)` ([docs](https://docs.python.org/3/library/weakref.html#weakref.finalize)) onto `cached.op()` which releases the physical cached copy from the databases.

The idempotency of `.cache()` is also preserved by returning the same table reference.

coroa commented:
The `bidi` dependency (for a bidirectional dictionary) is not necessary anymore with the RefCountedCache implementation in this PR, but quite obviously my [attempt](https://github.com/ibis-project/ibis/pull/9477/commits/980f37b99e166f70d8486fe241bf4a901ab5d7e9) at removing it was incomplete.

coroa commented on conda/environment.yml:
```diff
@@ -3,9 +3,8 @@ channels:
 dependencies:
   # runtime dependencies
-  - apache-flink
+  # - apache-flink
```
Ups, my custom environment fixes bled into this commit, inadvertently. Good thing, i had to revert it.

cpcloud commented on ibis/expr/types/relations.py:
```diff
@@ -4672,13 +4650,19 @@ def value_counts(self) -> ir.Table:
-    def release(self):
+    def release(self, stacklevel=2):
         """Release the underlying expression from the cache."""
+        warnings.warn(
```
Should we remove these in version 10? If so, let's add a message here about that.

Alternatively, I think `util.deprecated` might be possible to use here, which would give a slightly more standardized (standard from the perspective of Ibis anyway) error message.

jcrist replied:
I actually would prefer we keep these ways to explicitly release cached tables. Complex code can make understanding if the cached tables has any references left tricky, and can also result in code littered in dels (which is a code smell imo). There should be a way to make both patterns still work after this change.

cpcloud replied:
Yeah, that makes sense to me. Keeping these seems like a good idea.

coroa replied:
> There should be a way to make both patterns still work after this change.

Both patterns do work with the current implementation and can continue to work (the linear search in `release` should maybe be helped by an inverse dictionary).

coroa replied:
Do we want to remove the `DeprecationWarning` and maybe update the docs and release notes to make the new behaviour better known, then?

cpcloud replied:
Yeah, that sounds like a good idea!

coroa replied:
Ok, reverted the changes around the `DeprecationWarning` and added some additional detail to the docstring.

cpcloud commented on ibis/common/caching.py:
```diff
@@ -47,56 +49,50 @@ def __init__(
-            self.finalize(key)
+        cached = self.lookup(name)
+        finalizer = finalize(cached, self._release, key)
+        finalizer.atexit = False
```
Does this mean that cached tables are not guaranteed to be cleaned up?

coroa replied:
`atexit` means when called by the interpreter shutdown ([docs](https://docs.python.org/3/library/atexit.html)), ie. in a situation where the sessions are being closed (or have been, actually). I assumed that, since we create temporary tables, which afaik are only valid for a current session, it was redundant to explicitly request their removal.

coroa replied:
Do we want to remove the `DeprecationWarning` and maybe update the docs and release notes to make the new behaviour better known, then?

cpcloud replied:
It's not necessarily redundant, because unfortunately not all backends have support for temporary tables.

cpcloud replied:
I *think* we want keep the default value of `atexit` (`True`), and allow the cleanup to run on interpreter shutdown.

coroa replied:
> I think we want keep the default value of atexit (True), and allow the cleanup to run on interpreter shutdown.

Ok, then let's test that.

cpcloud replied:
> depends on whether the database connection is still there

True, and that's a big if since it renders the handler useless if the connection isn't there.

I'd say let's leave the default behavior for now.

cpcloud commented on ibis/common/caching.py:
```diff
@@ -29,8 +28,11 @@ def wrapper(*args, **kwargs):
+
 class RefCountedCache:
-    """A cache with reference-counted keys.
+    """A cache with (implicitly) reference-counted values.
```
```suggestion
    """A cache with implicitly reference-counted values.
```

Seems important enough to not parenthesize!

cpcloud commented on ibis/common/caching.py:
```diff
@@ -47,56 +49,50 @@ def __init__(
+        self.generate_name = generate_name
         self.key = key or (lambda x: x)
 
+        self.cache: dict[Any, CacheEntry] = dict()
```
Not to relitigate this whole PR but would a `WeakValueDictionary` also work here instead of explicit finalizers?

Not sure if `WeakValueDictionary` lets you control what happens on weakref finalization, but if it doesn't that I suppose that would be a non-starter.

cpcloud replied:
Seems like the finalizer behavior of the values in a `WeakValueDictionary` can't easily be customized.

coroa replied:
I actually had one originally and look into the implementation whether subclassing one could easily replace what i am doing here. I did not see a good way to do it more easily.

cpcloud commented on ibis/common/caching.py:
```diff
@@ -47,56 +49,50 @@ def __init__(
-        result = self.cache[key]
-        self.refs[key] += 1
-        return result
+        return self.cache[key].ref()
```
Should this raise a `KeyError` if it returns `None`?

coroa replied:
It needs to loose a race for that to happen (since the reference invalidation triggers the removal of the cache entry) and `__getitem__` is unused, but you are right!

cpcloud commented on ibis/common/caching.py:
```diff
@@ -47,56 +49,50 @@ def __init__(
+
+    def release(self, name: str) -> None:
+        # Could be sped up with an inverse dictionary,
+        # but explicit release is discouraged, anyway
```
This seems fine for now. Worst case scenario is enormous amounts of cached tables, but I suspect this would scale well into the tens of thousands. We can always revisit if this becomes a bottleneck.

cpcloud commented on ibis/expr/types/relations.py:
```diff
@@ -1881,8 +1882,8 @@ def difference(self, table: Table, *rest: Table, distinct: bool = True) -> Table
         node = ops.Difference(self, table, distinct=distinct)
-        for table in rest:
-            node = ops.Difference(node, table, distinct=distinct)
+        for other in rest:
+            node = ops.Difference(node, other, distinct=distinct)
```
Can we leave these as they were just to reduce review noise?

coroa replied:
pre-commit's ruff insisted. Happy to remove.

cpcloud reviewed (commented):
Thanks for digging around in the guts here :)

A couple of questions and requests, but I generally like where this PR is going!

cpcloud commented on ibis/common/caching.py:
```diff
@@ -47,56 +49,50 @@ def __init__(
+
+        return cached
+
+    def release(self, name: str) -> None:
```
Should we just merge the implementation of `_release` into this method?

cpcloud replied:
Oh, I see you're using it elsewhere, nevermind!

cpcloud commented on ibis/expr/types/relations.py:
```diff
@@ -4672,13 +4650,19 @@ def value_counts(self) -> ir.Table:
+    def release(self, stacklevel=2):
         """Release the underlying expression from the cache."""
+        warnings.warn(
+            "Explicitly releasing cached tables is deprecated "
+            "(they are released when the table relation leaves scope)",
```
```suggestion
            "Explicitly releasing cached tables is deprecated. "
            "Tables are released when the table expression leaves scope.",
```

cpcloud commented on ibis/common/caching.py:
```diff
@@ -47,56 +49,50 @@ def __init__(
         self.finalize = finalize
-        # Somehow mypy needs a type hint here
-        self.names: defaultdict = defaultdict(generate_name)
+        self.generate_name = generate_name
```
Seems like there's only a single call site for this, can we inline the `generate_name` in `store` and remove the attribute altogether?

cpcloud replied:
Can also be done in a follow-up!

cpcloud replied:
Actually, nevermind, I'll submit a PR to clean some stuff up here that's independent of this PR!

cpcloud commented:
I will clean up the `ruff` lints in #9489!

coroa commented:
> I will clean up the `ruff` lints in #9489!

Once that gets merged, i'll rebase ontop

cpcloud commented:
@coroa #9489 is merged!

cpcloud reviewed (approved):
LGTM!

coroa commented:
Ok, I rebased ontop of `main` and removed the DeprecationWarning and updated the `.cache()` doc-string with another sentence on the explicit and implicit cache eviction.

Only things left on my list of todos are:
- [ ] Release notes
- [ ] Check and update user guid

cpcloud commented:
Looks like there's maybe a bit of `DeprecationWarning` still hanging around?

cpcloud commented:
A release note will be generated from the commit message, so no need to do anything there.

coroa commented:
> Looks like there's maybe a bit of `DeprecationWarning` still hanging around?

Ah yes, i forgot about all the tests i had to change :).

coroa commented:
Couldn't find any user-guide explanations of the cache'ing mechanism. So i guess the doc-string is the best we have atm. And that is up-to-date.

So, from my side this is good to go.

cpcloud reviewed (approved):
LGTM!

@jcrist Did you wanna take a look here again?

jcrist commented on ibis/common/caching.py:
```diff
@@ -47,56 +49,53 @@ def __init__(
+
+    def _release(self, key) -> None:
+        entry = self.cache.pop(key)
+        self.finalize(entry.name)
```
This call may error for a few reasons:
- Table doesn't exist (I think we should ignore these always, as noted elsewhere).
- Spurious error
- Connection is in weird state since python is finalizing

Since _most_ backends will cleanup temp tables automatically (and errors on exit can give a bad impression of a tool), I wonder if want to do something like:

```python
try:
    self.finalize(entry.name)
except Exception:
    # silence all errors if system is shutting down
    if not sys.is_finalizing():
        raise
```

This way we can still catch bugs in our code since most release calls won't silence everything, but in the case the system is shutting down and e.g. the network is in a weird state we don't see a mountain of failed-to-release-table errors on exit.

coroa replied:
Ok, I added that.

jcrist reviewed (commented):
Overall LGTM! This is something we'd talked about doing internally before, nice to see it handled in such a nice way be a new contributor!

coroa reviewed (commented):
Thanks for the nice words

coroa reviewed (commented):
Thanks for the nice words

coroa commented on ibis/backends/sql/__init__.py:
```diff
@@ -275,8 +275,8 @@ def drop_view(
-    def _clean_up_cached_table(self, op):
-        self.drop_table(op.name)
+    def _clean_up_cached_table(self, name):
+        self.drop_table(name)
```
Done

coroa commented on ibis/backends/duckdb/__init__.py:
```diff
@@ -281,8 +281,8 @@ def create_table(
-    def _clean_up_cached_table(self, op):
-        self.drop_table(op.name)
+    def _clean_up_cached_table(self, name):
+        self.drop_table(name)
```
Done, and works

coroa commented on ibis/common/caching.py:
```diff
@@ -47,56 +49,53 @@ def __init__(
+        raise IbisError(
+            "Key has already been released. Did you call "
+            "`.release()` twice on the same expression?"
+        )
```
Makes sense

coroa commented:
Unsure what tripped the mssql backend. Unfortunately, have not been able to run those tests locally.

cpcloud commented:
One transient error and one XPASS, which is a win!

cpcloud commented:
I'll fix this up and merge!

cpcloud commented:
Clouds are passing:

```
…/ibis on  self-releasing-cached-tables is 📦 v9.1.0 via 🐍 v3.10.14 via ❄️   impure (ibis-3.10.14-env)
❯ pytest -m 'bigquery or snowflake' -n 8 --dist loadgroup --snapshot-update -q
bringing up nodes...
.x.x...........................x....s................x..........s...........................x.................xx..........x..x............x..........................x...x........................... [  5%]
............x....................x....................s..................s................................................s......................................................s..x.x.......x...... [ 10%]
x.......x.x..x.........x....x..x.xx....x..........x....................x....................................................x...............................................sssssssssssssssssssss.... [ 16%]
xxxxxxxxxx.x.xx.x.x....x.x.....................................x.............x........x......xx.x.........................x.......x....x.x...........x...x.......................x................... [ 21%]
...x....x..........................x.x....................x........x.......x.x.......x..x......................x..........................xx...............x......x.x..x...x...x......x.............. [ 26%]
.....x.............................x....xx.......x......x..xxxx.xx.x.x.xxx..xx.x...xxxx...x..xx.xxx..x.x..xx...xx.x.x...x......xxx.xxxx.xxx.x...xx.xx.xxxxxx.xx...x.xx.......x.........x.....x....... [ 32%]
xx..........x............................x...x..x..................x....x.....x............x.....x................x..x.....xxx..x.x...xxx.xxxx.xx......xx.x...x..x.....x..............x.............. [ 37%]
..x.x.x.x.............x..x.....x..x.x.......x.x............................x....x..x.x..x...x..x.....x...x..............x...xxxxx....x..................x.........x.......xx..x.x.................... [ 42%]
.....s.............x..............s.x...x....x..............sx..xx.s.x...x..x.......s......x.x......x..........x...................x...........x...........................x..................xx..... [ 48%]
.x.........x.....................x...........................s..........................s....................s..........x................................x..............x.x...x..x...........x...x... [ 53%]
..............xxx.....xx......................x.xx..x......x.x..x...x.x......x............x.x..x.......x..................x.......xx..xx..xx...xxxxxxxxxx...xxx...xxxx.xxxx..x...x.x..x..x.xxx.xx..x. [ 58%]
xxxxx..x.x.x...x..................x..x...x...x....x........x........x.....x.x....x.............x.x.x...xx.....x....x...x.......x...xxxxx................x................x..xx....xx................. [ 64%]
....x...x....x................x.............x.x.....x........x.x.x....................x..........x..xx.....x.......x.x...........x..xx.......x..x................x.....x......xxx...x.x.xxx..xxxxxxxx [ 69%]
xxxx.xxxxx.xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx..x......................x.........................x.....................x...x........ [ 74%]
....................................................................x.xxxxxx.xx.x....xx.xx....x..x.......x....................x..............x........x.x.xx...x.x...x.....x....x...xx..............x [ 80%]
..xx....x...xx..x..x.....x.....................x..x.................x.x.....x.......x...........x........x....x..x.x.......x.......x................................................................. [ 85%]
..........................................................................................................................................................................s.......................... [ 90%]
.........x.............................................................................................................................................................................x.x........... [ 96%]
......................................................s...ss......................................x...........................................                                                        [100%]
3097 passed, 39 skipped, 552 xfailed in 840.67s (0:14:00)
```

cpcloud commented:
@coroa Thanks for pushing this through, great work!

coroa commented:
Thanks for the quick reviews and comments!
